File: backend/app/services/arxiv.py
from __future__ import annotations

import re
import xml.etree.ElementTree as ET
from dataclasses import dataclass

import httpx

ARXIV_ID_RE = re.compile(
    r"(?:(?:https?://)?(?:www\.)?arxiv\.org/(?:abs|pdf)/)?"
    r"(?P<id>\d{4}\.\d{4,5}(?:v\d+)?|[a-z\-]+(?:\.[A-Z]{2})?/\d{7}(?:v\d+)?)",
    re.IGNORECASE,
)
ATOM_NS = {"atom": "http://www.w3.org/2005/Atom", "arxiv": "http://arxiv.org/schemas/atom"}


@dataclass
class ArxivPaper:
    arxiv_id: str
    title: str
    abstract: str | None
    authors: list[str]
    year: int | None
    pdf_url: str


def normalize_arxiv_id(value: str) -> str:
    raw = value.strip()
    match = ARXIV_ID_RE.search(raw)
    if not match:
        raise ValueError("Invalid arXiv id or URL")
    arxiv_id = match.group("id")
    # drop version suffix for storage uniqueness (keep pdf fetch with version if present)
    return arxiv_id


def canonical_arxiv_id(arxiv_id: str) -> str:
    return re.sub(r"v\d+$", "", arxiv_id, flags=re.IGNORECASE)


def fetch_arxiv_metadata(arxiv_id: str) -> ArxivPaper:
    query_id = arxiv_id
    url = f"https://export.arxiv.org/api/query?id_list={query_id}"
    with httpx.Client(timeout=30.0) as client:
        response = client.get(url)
        response.raise_for_status()

    root = ET.fromstring(response.text)
    entry = root.find("atom:entry", ATOM_NS)
    if entry is None:
        raise ValueError("arXiv paper not found")

    title = " ".join((entry.findtext("atom:title", default="", namespaces=ATOM_NS) or "").split())
    abstract = entry.findtext("atom:summary", default=None, namespaces=ATOM_NS)
    if abstract:
        abstract = " ".join(abstract.split())

    authors = [
        " ".join((node.findtext("atom:name", default="", namespaces=ATOM_NS) or "").split())
        for node in entry.findall("atom:author", ATOM_NS)
    ]
    authors = [name for name in authors if name]

    published = entry.findtext("atom:published", default=None, namespaces=ATOM_NS)
    year = int(published[:4]) if published and len(published) >= 4 else None

    pdf_url = f"https://arxiv.org/pdf/{canonical_arxiv_id(arxiv_id)}.pdf"
    for link in entry.findall("atom:link", ATOM_NS):
        if link.attrib.get("title") == "pdf" or link.attrib.get("type") == "application/pdf":
            href = link.attrib.get("href")
            if href:
                pdf_url = href
                break

    return ArxivPaper(
        arxiv_id=canonical_arxiv_id(arxiv_id),
        title=title or f"arXiv:{canonical_arxiv_id(arxiv_id)}",
        abstract=abstract,
        authors=authors,
        year=year,
        pdf_url=pdf_url,
    )


def download_pdf(pdf_url: str) -> bytes:
    with httpx.Client(timeout=60.0, follow_redirects=True) as client:
        response = client.get(pdf_url)
        response.raise_for_status()
        return response.content

File: backend-go/internal/services/arxiv.go
package services

import (
	"context"
	"encoding/xml"
	"fmt"
	"io"
	"net/http"
	"regexp"
	"strings"
)

var arxivRE = regexp.MustCompile(`(?i)(?:https?://(?:www\.)?arxiv\.org/(?:abs|pdf)/)?(\d{4}\.\d{4,5}(?:v\d+)?|[a-z-]+(?:\.[A-Z]{2})?/\d{7}(?:v\d+)?)`)

type ArxivPaper struct {
	ArxivID, Title, Abstract, PDFURL string
	Authors                          []string
	Year                             *int
}
type atomFeed struct {
	Entries []atomEntry `xml:"entry"`
}
type atomEntry struct {
	Title     string `xml:"title"`
	Summary   string `xml:"summary"`
	Published string `xml:"published"`
	Authors   []struct {
		Name string `xml:"name"`
	} `xml:"author"`
	Links []struct {
		Href  string `xml:"href,attr"`
		Type  string `xml:"type,attr"`
		Title string `xml:"title,attr"`
	} `xml:"link"`
}

func NormalizeArxivID(value string) (string, error) {
	m := arxivRE.FindStringSubmatch(strings.TrimSpace(value))
	if len(m) < 2 {
		return "", fmt.Errorf("Invalid arXiv id or URL")
	}
	return m[1], nil
}
func CanonicalArxivID(id string) string {
	return regexp.MustCompile(`(?i)v\d+$`).ReplaceAllString(id, "")
}
func FetchArxivMetadata(ctx context.Context, id string) (ArxivPaper, error) {
	req, err := http.NewRequestWithContext(ctx, http.MethodGet, "https://export.arxiv.org/api/query?id_list="+id, nil)
	if err != nil {
		return ArxivPaper{}, err
	}
	resp, err := http.DefaultClient.Do(req)
	if err != nil {
		return ArxivPaper{}, err
	}
	defer resp.Body.Close()
	if resp.StatusCode/100 != 2 {
		return ArxivPaper{}, fmt.Errorf("arXiv returned %s", resp.Status)
	}
	var feed atomFeed
	if err = xml.NewDecoder(resp.Body).Decode(&feed); err != nil {
		return ArxivPaper{}, err
	}
	if len(feed.Entries) == 0 {
		return ArxivPaper{}, fmt.Errorf("arXiv paper not found")
	}
	e := feed.Entries[0]
	p := ArxivPaper{ArxivID: CanonicalArxivID(id), Title: strings.Join(strings.Fields(e.Title), " "), Abstract: strings.Join(strings.Fields(e.Summary), " "), PDFURL: "https://arxiv.org/pdf/" + CanonicalArxivID(id) + ".pdf"}
	if p.Title == "" {
		p.Title = "arXiv:" + p.ArxivID
	}
	if len(e.Published) >= 4 {
		var y int
		if _, err = fmt.Sscanf(e.Published[:4], "%d", &y); err == nil {
			p.Year = &y
		}
	}
	for _, a := range e.Authors {
		if n := strings.Join(strings.Fields(a.Name), " "); n != "" {
			p.Authors = append(p.Authors, n)
		}
	}
	for _, l := range e.Links {
		if l.Title == "pdf" || l.Type == "application/pdf" {
			p.PDFURL = l.Href
			break
		}
	}
	return p, nil
}
func DownloadPDF(ctx context.Context, url string) ([]byte, error) {
	req, err := http.NewRequestWithContext(ctx, http.MethodGet, url, nil)
	if err != nil {
		return nil, err
	}
	resp, err := http.DefaultClient.Do(req)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()
	if resp.StatusCode/100 != 2 {
		return nil, fmt.Errorf("PDF download returned %s", resp.Status)
	}
	return io.ReadAll(io.LimitReader(resp.Body, 51<<20))
}
